src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,74 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) {
+    return Result();
+  }
   // Invoked after all the children of `module` is visited.
-  virtual void PostVisit(const Module& module) {}
+  virtual Result PostVisit(const Module& module, Result result) {
+    return result;
```

Comment:
Does this need to be `std::move(result)` if we want to take advantage of zero copy? @markww @Cypher1 

(This is one of the reasons why I don't quite like `std::move`. There is too much of a cognitive overload that we have to be careful everywhere to make sure we have zero copies. )